Custom readline extension

georgebrock opened the pull request:
This is still work in progress, but it's got pretty big and pretty complex.

Details of what this is, what's done, and what's still to do are on issue #255.

mike-burns commented on Makefile.am:
```diff
@@ -1,5 +1,5 @@
 AUTOMAKE_OPTIONS = foreign
-SUBDIRS = src man lib/gitsh vendor spec
+SUBDIRS = src man vendor spec ext/gitsh lib/gitsh
```
Does this order matter? Or: was this a technical change that I should learn about?

georgebrock replied:
I don't believe this matters. Not sure why I changed it now (I did the initial setup for this branch more than a year ago)

mike-burns commented on ext/gitsh/src/line_editor.c:
```diff
@@ -0,0 +1,1973 @@
+/************************************************
```
Does this file differ from upstream? Is it possible to clarify which is copied and which isn't? Or perhaps use e.g. `#include` to separate the files?

georgebrock replied:
ef19570 introduced an exact copy of the upstream module, and subsequent commits made changes to it.

You can see just the changes here: https://github.com/thoughtbot/gitsh/compare/ef19570...gb-custom-readline-extension#diff-9

mike-burns commented on spec/units/line_editor_blank_filter_spec.rb:
```diff
@@ -0,0 +1,56 @@
+require 'spec_helper'
```
Great pattern you've set up in this file -- easy to read.

georgebrock replied:
Credit where it's due: I just renamed this file, it was @sgrif and @adam89 who first wrote it.

mike-burns commented on spec/units/line_editor_spec.rb:
```diff
@@ -0,0 +1,702 @@
+require 'spec_helper'
```
This file is exciting. Not sure I understand all of it on a first read. Anything in particular I should focus on?

georgebrock replied:
It's mostly a port of the Ruby test suite's Readline tests, converted from minitest to RSpec, and better organised by method.

I intend to do more refactoring here before this is merged.

mike-burns commented on ext/gitsh/src/line_editor.c:
```diff
@@ -0,0 +1,1973 @@
+        rb_obj_hide(str);
+#else
+        RBASIC(str)->klass = 0;
+#endif
```
Maybe this could do the same preprocessor trick as `readline_s_delete_bytes`:

```c
#if !defined(HAVE_RB_OBJ_HIDE)
#define rb_obj_hide(str) do{ RBASIC(str)->klass = 0; } while(0);
#end
```

(Same with `rb_obj_reveal` below.)

Then you won't need to modify these functions directly.

georgebrock replied:
Could you explain what the `do{ … } while(0);` is for? I've not seen that before.

mike-burns replied:
In the exact case that you showed, it will work fine without the `do/while`. However, C macros break in cases like this:

```c
#define DO_THE_THING(x) thing1(x); thing2(x);
if (true)
        DO_THE_THING(str);
else
        warnx("no thing found");
```

So the `do/while` was a habit and not technically needed in this case: this case is where the macro expands to exactly one statement, and it is not used inside a branch of a braceless block.

georgebrock replied:
Makes sense, thanks.

georgebrock commented:
@mike-burns Thanks for the feedback so far. I think this is now ready to go.

Since you last looked, I've:

* Dropped support for libedit,
* dropped support for a bunch of less useful readline features,
* refactored the `Gitsh::LineEditor` specs so that they're somewhat readable,
* used macros in C code in the ways you suggested, and
* updated the documentation.

This PR appears to solve #114. It also sets the stage for solving #150 and #189, but doesn't take  the crucial last step of exposing the `rl_char_is_quoted_p` setting; there's already enough going on here without adding new functionality, so I'll save that for another PR.

mike-burns commented on CONTRIBUTING.md:
```diff
@@ -13,15 +13,33 @@ Here's a quick guide to begin contributing:
+
+        brew install readline
+
 2. Build the generated files. Some Ruby files in `gitsh` are generated by the
```
Want to update this `2` to a `3`, and so on?

georgebrock replied:
Good catch.

mike-burns commented on ext/gitsh/src/line_editor.c:
```diff
@@ -0,0 +1,1764 @@
+/************************************************
+
+  Based on Ruby 2.3.0's
```
:+1:

mike-burns commented on lib/gitsh/line_editor.rb:
```diff
@@ -0,0 +1,31 @@
+      byte_beg = line_buffer[0...beg].bytesize
+      byte_len = line_buffer[beg, len].bytesize
+
+      self.delete_bytes(byte_beg, byte_len)
```
:tada:

georgebrock replied:
This was much easier than getting the C implementation of `delete_text` to work properly across Ruby version with multi-byte strings!

mike-burns commented:
Excellent!

Do you know: does this behave well on a Ruby that was built against libedit? My understanding is that it will work fine: you're linking against Readline and ignoring whatever Ruby does.

LGTM.

georgebrock commented on ext/gitsh/src/line_editor.c:
```diff
@@ -0,0 +1,1764 @@
+
+
+/*
+ * Document-class: Readline
```
I should probably change the documentation in this file: it refers to the module as `Readline` (it's now `Gitsh::LineEditor`) and makes reference to supporting libedit (which we no longer support).

mike-burns replied:
Good catch. Is it worth it to even leave this documentation here?

georgebrock replied:
Deleting it is definitely one way to update it!

georgebrock commented:
> Do you know: does this behave well on a Ruby that was built against libedit? My understanding is that it will work fine: you're linking against Readline and ignoring whatever Ruby does.

That's definitely the intention, but I don't think I've explicitly tested it. I will do so, and report back.

georgebrock commented:
Yep, this works fine with a Ruby version built against libedit.

georgebrock commented:
Given the high complexity of this, and how many related things there are, I decided to keep a larger number of distinct commits than I would for a typical PR, and to merge them with `--no-ff` to keep them grouped together in the Git history.
